docs/design.md

```diff
@@ -109,11 +109,19 @@ A fully-populated example of an `UpgradeConfig` status is included below:
 
 ## Config Managers
 
-The `managed-upgrade-operator` provides a configurable mechanism for retrieving and storing an `UpgradeConfig` 
+The `managed-upgrade-operator` provides a configurable mechanism for retrieving and storing an `UpgradeConfig`
 Custom Resource that can be reconciled against by the `UpgradeConfig` controller.
 
 For more information, see the dedicated section on this topic: [UpgradeConfig Managers](configmanager.md)
 
+## Controllers
+
+The `managed upgrade operator` provided upgrade process revolves around multiple Controllers. Alongside the above mentioned `UpgradeConfig` controller, the `NodeKeeper` controller works simultaneously in an upgrade process towards the state of nodes in the cluster.
+
+The `NodeKeeper` controller keeps a track of state of all the nodes during an upgrade, making sure of any activities around the nodes and acts upon any action required towards maintaining the required state of any node.
+
+If a node is having any troubles while an ongoing upgrade, the `NodeKeeper` controller will perform necessary action towards the state of the node and upgrade continuation.
```

Comment:
```suggestion
If an upgrading worker node is experiencing difficulty draining due to conditions such as [Pod Disruption Budgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) or stuck finalizers), the `NodeKeeper` controller will perform remediation strategies to ensure the node's eventual drain and subsequent upgrade continuation.

The `NodeKeeper` controller will flag through metrics any worker node that continue to unsuccessfully drain in spite of the remediation strategies. 
```

Reply:
Thanks for the suggestion, I have made changes according to the suggestion.